### ISSUE TITLE
[lsp server] column offsets don't follow LSP specification mandated UTF-16, creating problems in emacs client

### ISSUE THREAD

**Issue by Rehan-MALAK** [Deducteam/lambdapi#452]
@amelieled 
Hi ! Do you use emacs or vscode ?
Have you been able to use emacs lsp mode on tests/OK/inductive_type.lp
(it's freezing on my laptop)
thanks

can someone reproduce this behaviour or is it just me ?

PS : not related to #451 as neither 9e080cd7 4a54066f 8c5755f6 74de6107 seem to work

**Comment by ejgallego** [comment]
Seems pretty sure to be a duplicate of #451

**Comment by Rehan-MALAK** [comment]
#453 fixed tests/OK/unif_hint.lp #451 on my laptop

but I still can't use tests/OK/inductive_type.lp (one of the CPU is at 100% and have to kill emacs)

can someone reproduce this behaviour or is it just me ?

**Comment by amelieled** [comment]
Hi @Rehan-MALAK 
You speak about the current version of the code, or about a precise PR?

**Comment by amelieled** [comment]
With the current version, I haven't problem (except that I should use the command `opam install alcotest` before the command `make` @gabrielhdt )

**Comment by Rehan-MALAK** [comment]
> #453 fixed tests/OK/unif_hint.lp #451 on my laptop
> 
> but I still can't use tests/OK/inductive_type.lp (one of the CPU is at 100% and have to kill emacs)
> 
> can someone reproduce this behaviour or is it just me ?

I have written the precise commits where I can't make it work. But if you say that emacs+lsp+inductive_type works on your side, I have to investigate more

**Comment by Rehan-MALAK** [comment]
I don't think this issue should be closed. 

```lambdapi
constant symbol Prop : TYPE
injective symbol P : Prop → TYPE
set builtin "Prop" ≔ Prop
set builtin "P"    ≔ P 

inductive R : TYPE ≔
 | i  : R

set declared "ℝ"
constant symbol ℝ : TYPE
constant symbol j : ℝ 

set declared "𝔹"
inductive 𝔹 : TYPE ≔
```
but if I add a constructor to 𝔹
```
| k : 𝔹
```

I have to kill emacs

**Comment by amelieled** [comment]
Yes, there is a problem with the special string like 𝔹.
It's due to the parser, I think.

**Comment by ejgallego** [comment]
I cannot reproduce, are you sure you are testing on master?

What does the log in `/tmp/lambdapi_lsp_log.txt` say?

**Comment by Rehan-MALAK** [comment]
Yes this is master ...

@amelieled or somebody else
can you please tell us if you can reproduce the bug with current master 1068b496 and emacs+lambdapi mode

no problem with this :
```
constant symbol Prop : TYPE
injective symbol P : Prop → TYPE
set builtin "Prop" ≔ Prop
set builtin "P"    ≔ P 
set declared "𝔹"
inductive 𝔹 : TYPE ≔ 
| k :  𝔹
```

and emacs freezing with this :
```
constant symbol Prop : TYPE
injective symbol P : Prop → TYPE
set builtin "Prop" ≔ Prop
set builtin "P"    ≔ P 
set declared "𝔹"
inductive 𝔹 : TYPE ≔ | k :  𝔹
```

**Comment by fblanqui** [comment]
It works well for me:
emacs 26.1
eglot 1.6
ocaml 4.07.1

**Comment by ejgallego** [comment]
Failed at reproducing again, @Rehan-MALAK please attach the full log file and I will have a look.

**Comment by Rehan-MALAK** [comment]
> Failed at reproducing again, @Rehan-MALAK please attach the full log file and I will have a look.

I posted /tmp/lambdapi_lsp_log.txt in the previous comment. I would like to help but I don't know which full log you are refering to.

**Comment by ejgallego** [comment]
Oh, sorry @Rehan-MALAK I missed it :/ , reading too fast I guess :)

So indeed that's a strange one:
```
[[fatal error]]: End_of_file
[[BT]]: Raised at file "stdlib.ml", line 441, characters 14-31
Called from file "src/lsp/lsp_io.ml", line 27, characters 11-24
Called from file "src/lsp/lp_lsp.ml", line 432, characters 14-36
Called from file "src/lsp/lp_lsp.ml", line 438, characters 6-13
```
which is:
```
let read_request ic =
  let cl = input_line ic in
  let sin = Scanf.Scanning.from_string cl in
  try
    let raw_obj =
      Scanf.bscanf sin "Content-Length: %d\r" (fun size ->
```
so indeed the first `input_line` is throwing `End_of_file`, that's bizarre! What is emacs sending here?

If my interpretation of the protocol is correct the `Content-Length` header is always sent. Is maybe the lsp fifo closed on the emacs side? Can you reproduce with a bare-bones emacs setup?

**Comment by ejgallego** [comment]
For reference ocaml documentation says that `input_line` will raise `End_of_file` indeed only when the input channel reaches EOF at the beggining of the line.

**Comment by Rehan-MALAK** [comment]
Sorry I dont understand how I have done with the previous attached one... maybe it was after killing emacs...
[lambdapi_lsp_log.txt](https://github.com/Deducteam/lambdapi/files/5466108/lambdapi_lsp_log.txt)

**Comment by ejgallego** [comment]
Thanks for the update.

So indeed the problem seems like our lsp server is sending a position `{ l : 6, c : 8}` that indeed goes beyond the end of the buffer. This is because we were lazy [and didn't link enough dependencies] so we count columns in 8-byte chars, not in utf-8 columns.

Couple of observations:
- it is definitively an emacs bug that it hangs if the server sends an invalid location, it should just error / warn / ignore. This should be reported emacs upstream
- we ought to fix our server internals w.r.t. positions; the LSP standard [specifies](https://microsoft.github.io/language-server-protocol/specification#textDocuments) that buffer text positions are to be reported relative to utf-16.

As of today, the LSP server uses the positions that LP parser reports; indeed the best would be to update the parser so it can handle utf-16 relative positions properly but this is far from a trivial endeavor. Doing the conversion in the server is also quite a pain as it would require re-interpreting every line and doing a position-based lookup map.

What do you think folks, how should we address this issue?

Meanwhile, I guess that adding some extra comment at the end of the file is a workaround until we fix this.

**Comment by Rehan-MALAK** [comment]
I am not sure what you mean... If I do
```
constant symbol Prop : TYPE
injective symbol P : Prop → TYPE
set builtin "Prop" ≔ Prop
set builtin "P"    ≔ P 
set declared "𝔹"
inductive 𝔹 : TYPE ≔
| k :  𝔹
// comment
```
or
```
constant symbol Prop : TYPE
injective symbol P : Prop → TYPE
set builtin "Prop" ≔ Prop
set builtin "P"    ≔ P 
set declared "𝔹"
inductive 𝔹 : TYPE ≔
| k :  𝔹 // comment
```
I have the same problem 
```
              {
                "range": {
                  "start": { "line": 5, "character": 0 },
                  "end": { "line": 6, "character": 8 }
                },
                "severity": 4,
                "message": "OK"
              }
```

**Comment by ejgallego** [comment]
I dunno then, the problem seems clear in that emacs is trying to access an invalid column; if you replace the unicode `𝔹` for `bool` does the problem go away?

**Comment by fblanqui** [comment]
@gabrielhdt will the new parser (https://github.com/Deducteam/lambdapi/pull/441) able to handle utf16 characters?

**Comment by gabrielhdt** [comment]
It can since sedlex handles utf16: https://ocaml-community.github.io/sedlex/sedlex/Sedlexing/Utf16/index.html

**Comment by Rehan-MALAK** [comment]
So , for reference

"Double-Struck Capital B" UTF8 **4 bytes** 0xF0 0x9D 0x94 0xB9, unicode U+1D539 https://www.compart.com/fr/unicode/U+1D539    => KO

```
set declared "𝔹"
constant symbol 𝔹 : TYPE
constant symbol j : 𝔹
```

"Double-Struck Capital R" UTF8 **3 bytes** 0xE2 0x84 0x9D,  unicode U+211D https://www.compart.com/fr/unicode/U+211D      => OK
```
set declared "ℝ"
constant symbol ℝ : TYPE
constant symbol j : ℝ
```

Do these informations confirm what you deduced ?

**Comment by ejgallego** [comment]
> At the end, have you been able to reproduce the error : the fact that emacs become unusable.

No, but likely due to my emacs / eglot version?

**Comment by ejgallego** [comment]
> @gabrielhdt will the new parser (#441) able to handle utf16 characters?

Note that I'm unsure if the right solution if for `lambdapi` to switch to utf-16, I was reading a bit about this choice in LSP and indeed it is quite a mess; in particular documents are sent in utf-8 but positions are in utf-16.

The situation is quite puzzling and I'm unsure what to do. I'd say that lambdapi should use the format that better works for it, then we will have the LSP layer to do the conversions needed.

p.s: I believe the reason they chose UTF-16 for positions is that Javascript is supposed to store strings in that encoding, but indeed that has created many problems for the rest of the world.

**Comment by gabrielhdt** [comment]
All right, with the new parser, hopefully it's just a switch in the lexer (using the utf16 module instead of utf8).

**Comment by ejgallego** [comment]
> All right, with the new parser, hopefully it's just a switch in the lexer (using the utf16 module instead of utf8).

That's good news in terms of reducing the hacks; still LSP sends the documents in UTF-8 but that's easier to deal with [conversion in input / output]

@Rehan-MALAK , I was discussing with the ocaml-lsp people to see if they can provide a library version to help with this [as likely they have the same problem]

It is gonna be a few days for sure before we can implement a workaround, so what I'd suggest here is to report the problem in the emacs side too [emacs should not hang if the server provides a buggy position] , WDYT?

**Comment by Rehan-MALAK** [comment]
> report the problem in the emacs side

yes I have to do that
I should learn first how to trigger this bug wihout installing eglot, ocaml and lambdapi
then submit it accordingly to https://www.gnu.org/software/emacs/manual/html_node/efaq/Reporting-bugs.html

**Comment by ejgallego** [comment]
By the way, we have been discussing about the best way to fix this [as all this translation are quite involved]

cc: https://github.com/ocaml/ocaml-lsp/issues/304

**Comment by ejgallego** [comment]
> I should learn first how to trigger this bug wihout installing eglot, ocaml and lambdapi

That seems like a bug in eglot, so I think that may be hard to reproduce it without installing?

**Comment by Rehan-MALAK** [comment]
Well , you seemed confident at first that it was in emacs side...
> it is definitively an emacs bug


> I'd suggest here is to report the problem in the emacs side too

It turned out that it is not a bug, neither in emacs, neither in eglot ... but an eglot misconfiguration problem... We can choose the move-to-column function depending on if the server is compliant or not with the spec...

https://github.com/joaotavora/eglot/blob/master/eglot.el#L1104-L1115

If I put this in my ~/.emacs, 
```lisp
(setq eglot-move-to-column-function 'eglot-move-to-column)
```
the problem disappear

> # 1.6 


> ##### Column offset calculation is now LSP-conform ([#361][github#361])
> 
> It seems the majority of servers now comply with the language server
> specification when it comes to handling non-ASCII texts.  Therefore
> the default values of `eglot-move-to-column-function` and
> `eglot-current-column-function` have been changed.  The documentations
> of these variables help to restore the old behavior.

https://github.com/joaotavora/eglot/commit/cacfd2410872d1b3895e8fef5a8068937f1f857d

**Comment by ejgallego** [comment]
> Well , you seemed confident at first that it was in emacs side...

Umm, where did I say that? The bug is in lambdapi for sure, however emacs should not hang on incorrect messages from the server, IMHO `eglot-move-to-lsp-abiding-column` should be made more resilient.

It is good to know there is a workaround already! Indeed this is a widespread issue among servers.

**Comment by fblanqui** [comment]
To conclude: we just have to add in the LP documentation that Emacs users should add `(setq eglot-move-to-column-function 'eglot-move-to-column)` in their `~/.emacs`. Isn't it possible to add this in the Emacs LP mode file so that the user doesn't have to do it? @firewall2142 could you please do this if that is possible, or else update the doc, so that we can close this issue?

**Comment by ejgallego** [comment]
Yup, the setup should be done.

> so that we can close this issue?

Unfortunately the above is a workaround, the issue still exists until the LSP specification is fixed. So to summarize on what remains to be done here:
- we need to implement a bidirectional conversion from UTF-8 <-> UTF-16 for the document
- moreover, we need to translate positions in the document in this way, that's the main tricky point!
- once that is done, we need to hook the translation in a few places:
  + documentChanged: when supporting diffs, start/end positions are in utf-16 [tho text is in UTF-8, what a mess LSP]
  + diagnostics need to be translated from UTF-8 to UTF-16

Alternatively, we could have the internal format of lambdapi to be UTF-16, but that's far from trivial, as it would mean replacing `String` by `UString` in the whole codebase [tho, it could be a good idea in the long term to have something like this]

**Comment by ejgallego** [comment]
reopening as 9a19146e9f3c6192216da589447c2c9fd3c2cc76 is merely a work-around, we should document the current limitation here for now. Hopefully we can take code from ocaml-lsp to solve this.

**Comment by fblanqui** [comment]
It seems that we can close this issue now that #977 is merged.
